exchanges/ftx/ftx_websocket.go

```diff
@@ -0,0 +1,485 @@
+		sub.Channel = channelToSubscribe.Channel
+	default:
+		sub.Operation = subscribe
+		sub.Channel = channelToSubscribe.Channel
+		sub.Market = f.FormatExchangeCurrency(channelToSubscribe.Currency, a).String()
+	}
+	return f.WebsocketConn.SendJSONMessage(sub)
+}
+
+// GenerateDefaultSubscriptions generates default subscription
+func (f *FTX) GenerateDefaultSubscriptions() {
+	var channels = []string{wsTicker, wsTrades, wsOrderbook, wsMarkets, wsFills, wsOrders}
+	var subscriptions []wshandler.WebsocketChannelSubscription
+	for a := range f.CurrencyPairs.AssetTypes {
+		pairs := f.GetEnabledPairs(f.CurrencyPairs.AssetTypes[a])
+		for z := range pairs {
+			newPair := currency.NewPairWithDelimiter(pairs[z].Base.String(), pairs[z].Quote.String(), "-")
+			for x := range channels {
+				subscriptions = append(subscriptions, wshandler.WebsocketChannelSubscription{
+					Channel:  channels[x],
+					Currency: newPair,
+				})
+			}
+		}
+	}
+	f.Websocket.SubscribeToChannels(subscriptions)
+}
+
+// wsReadData gets and passes on websocket messages for processing
+func (f *FTX) wsReadData() {
+	f.Websocket.Wg.Add(1)
+
+	defer f.Websocket.Wg.Done()
+
+	for {
+		select {
+		case <-f.Websocket.ShutdownC:
+			return
+
+		default:
+			resp, err := f.WebsocketConn.ReadMessage()
+			if err != nil {
+				f.Websocket.ReadMessageErrors <- err
+				return
+			}
+			f.Websocket.TrafficAlert <- struct{}{}
+			err = f.wsHandleData(resp.Raw)
+			if err != nil {
+				f.Websocket.DataHandler <- err
+			}
+		}
+	}
+}
+
+func timestampFromFloat64(ts float64) time.Time {
+	secs := int64(ts)
+	nsecs := int64((ts - float64(secs)) * 1e9)
+	return time.Unix(secs, nsecs)
+}
+
+func (f *FTX) wsHandleData(respRaw []byte) error {
+	var result map[string]interface{}
+	err := json.Unmarshal(respRaw, &result)
+	if err != nil {
+		return err
+	}
+	switch result["type"] {
+	case wsUpdate:
+		var p currency.Pair
+		var a asset.Item
+		market, ok := result["market"]
+		if ok {
+			p = currency.NewPairFromString(market.(string))
+			a, err = f.GetPairAssetType(p)
+			if err != nil {
+				return err
+			}
+		}
+		switch result["channel"] {
+		case wsTicker:
+			var resultData WsTickerDataStore
+			err = json.Unmarshal(respRaw, &resultData)
+			if err != nil {
+				return err
+			}
+			f.Websocket.DataHandler <- &ticker.Price{
+				ExchangeName: f.Name,
+				Bid:          resultData.Ticker.Bid,
+				Ask:          resultData.Ticker.Ask,
+				Last:         resultData.Ticker.Last,
+				LastUpdated:  timestampFromFloat64(resultData.Ticker.Time),
+				Pair:         p,
+				AssetType:    a,
+			}
+		case wsOrderbook:
+			var resultData WsOrderbookDataStore
+			err = json.Unmarshal(respRaw, &resultData)
+			if err != nil {
+				return err
+			}
+			if len(resultData.OBData.Asks) == 0 && len(resultData.OBData.Bids) == 0 {
+				return nil
+			}
+			err = f.WsProcessUpdateOB(&resultData.OBData, p, a)
+			if err != nil {
+				f.wsResubToOB(p)
+				return err
+			}
+		case wsTrades:
```

Comment:
I am not getting trades come through. 

```
{"channel": "trades", "market": "BTC/USD", "type": "update", "data": [{"id": 48365748, "price": 9540.0, "size": 0.004, "side": "sell", "liquidation": false, "time": "2020-06-01T05:53:14.048363+00:00"}, {"id": 48365749, "price": 9540.0, "size": 0.0002, "side": "sell", "liquidation": false, "time": "2020-06-01T05:53:14.048363+00:00"}, {"id": 48365750, "price": 9536.5, "size": 0.0211, "side": "sell", "liquidation": false, "time": "2020-06-01T05:53:14.048363+00:00"}]}
```

Is not reflected on the data handler side.